### PR TITLE
iOS learning group session four materials

### PR DESCRIPTION
**Pull request by ashfurrow**
Proposed materials for tomorrow's class.

### CODE REVIEW

File: resources/mobile/learning-group/session-four.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+- What is the purpose of a view controller, from a developer's perspective? What about from a user's perspective?
+
+<details>
+<summary>Answer</summary>
+
+A view controller is the main unit of composition for building native iOS UI's. From a user's perspective, they
+might call it "a screen", since view controllers commonly take up the entire screen of an app. However, view
+controllers really can be _composed_ so while a view controller might indeed take up the entire screen, it might be
+composed of child controllers as the develop sees fit.
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]: I feel like this is missing some explanation about what a view controller is conceptually responsible for, in MVC.

File: resources/mobile/learning-group/session-four.md [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+</details>
+
+- How does a parameters on the Objective-C view controller `init` method get sent to the React component hierarchy
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]: This sentence doesn’t read right.

File: resources/mobile/learning-group/session-four.md [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+</details>
+
+- We've seen how to create new controllers and link them with Eigen. Routing between view controller is all
+  URL-based; what are the advantages and disadvantages to this approach?
+
+<details>
+<summary>Answer</summary>
+
+**Advantages**: Getting to use the same URL's as the web means push notifications, deep links, and email deep links
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
**Advantages**: Getting to use the same URLs as the web means push notifications, deep links, and email deep links
```

File: resources/mobile/learning-group/session-four.md [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+all use the same routing infrastructure as UI interactions.
+
+**Disadvantages**: Data passed from one controller to another has to be serialized into a URL first. Common
+patterns such as passing `delegate` references or completion handlers forward to new view controllers is not
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]: This sentence doesn’t read right. Is it supposed to be the following?

```suggestion
patterns such as passing `delegate` references or forwarding completion handlers to new view controllers is not
```

File: resources/mobile/learning-group/session-four.md [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+possible.
+
+</details>
+
+- Related: in `ARSwitchBoard.m`, you'll see two different methods for adding routes: `addRoute:handler:` and
+  `registerEchoRouteForKey:handler:`. This lets us define routing URL's in [Echo](https://github.com/artsy/echo),
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
  `registerEchoRouteForKey:handler:`. This lets us define routing URLs in [Echo](https://github.com/artsy/echo),
```

File: resources/mobile/learning-group/session-four.md [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+  Artsy's feature-flag-as-a-service. Why do you think this added functionality here?
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
  Artsy's feature-flag-as-a-service. Why do you think this functionality was added here?
```

File: resources/mobile/learning-group/session-four.md [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+</details>
+
+- Extra credit: Emission's `ARComponentViewController` defines a few props that get passed down based on iOS View
+  Controller lifecycle methods. What are they? (Hint: [here is a PR](https://github.com/artsy/emission/pull/1890)
+  that uses them.)
+
+<details>
+<summary>Answer</summary>
+
+The `isVisible` is passed to all top-level components (ie: components backed by a component view controller). This
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
The `isVisible` prop is passed to all top-level components (ie: components backed by a component view controller). This
```

File: resources/mobile/learning-group/session-four.md [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+fetches actual data from Metaphysics v2. To keep things simple, **we will be staying totally within the Emission
+test app**.
+
+Let's start by checking out the code from the end of last session:
+
+```sh
+# cd into Emission's repo
+git fetch --tags # Required!
+git checkout ios-learning-session-three-finished
+git checkout -b ios-learning-group-session-four
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]: If you want, you can collapse the last two lines into:

```
git checkout -b ios-learning-group-session-four ios-learning-session-three-finished
```

File: resources/mobile/learning-group/session-four.md [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
++  return <MyNewComponent artistID="tim-fishlock" />
+ })
+```
+
+And that's it! Completed code from this session can be found in Emission's repo at
+[the `ios-learning-group-session-four-finished` tag](https://github.com/artsy/emission/releases/tag/ios-learning-group-session-four-finished).
+
+## What about the Query Map?
+
+When ran in development mode, Emission sends the full GraphQL query and variables to Metaphysics. However, in
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]: ?

```suggestion
When ran in development mode, Emission sends the full GraphQL query and variables to Metaphysics. However,
```

File: resources/mobile/learning-group/session-four.md [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+deployed versions (App Store releases **and** betas!) don't do this. Instead, they use GraphQL persisted queries.
+Metaphysics has a
+[flat JSON file](https://github.com/artsy/metaphysics/blob/9b782b05e9e7da08937581fddb4ef1cf315e2136/src/data/complete.queryMap.json)
+that defines every query of every version of Emission ever released. When a new version of Emission is made, the
+release script automatically generates a new query map, merges it with the existing one, and opens a pull request
+([exmaple](https://github.com/artsy/metaphysics/pull/1993)).
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
([example](https://github.com/artsy/metaphysics/pull/1993)).
```

File: resources/mobile/learning-group/session-four.md [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+## Core Concept Review & Homework
+
+- Come prepared with a bug or feature from your team's backlog to work on. Session Five will be a working session,
+  where we'll either pair or mob as a group on these issues.
+- We saw how React apps, and Emission in particular, avoid DRY software development. What do you think of this?
+  Why?
+- Emission is currently the only client to use persisted queries. Why do you think Emission uses them while other
+  client apps (like Volt and Reaction) don't yet?
+- Extra credit: We see that the content of our component is touching the top edge of the simulator screen. This
+  makes sense,but poses a problem. What do you think the problem is, and how do you think we could fix it? Hint:
```
Comment (alloy) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
  makes sense, but poses a problem. What do you think the problem is, and how do you think we could fix it? Hint:
```

File: resources/mobile/learning-group/session-four.md [hunk 12]
```diff
@@ -0,0 +1,363 @@
+# iOS Learning Group Session Three
```
Comment (anandaroop) [inline resources/mobile/learning-group/session-four.md]:
```suggestion
# iOS Learning Group Session Four
```

**Pull request merged**